src/util/openfaas/index.js

```diff
@@ -0,0 +1,149 @@
+const NodeCache = require("node-cache");
+const {
+  getFunction,
+  deleteFunction,
+  deployFunction,
+  invokeFunction
+} = require("./faasApi");
+const logger = require("../../logger");
+const { RetryRequestError } = require("../utils");
+
+const FAAS_BASE_IMG =
+  process.env.FAAS_BASE_IMG || "rudderlabs/openfaas-flask:main";
+const FAAS_MAX_PODS_IN_TEXT = process.env.FAAS_MAX_PODS_IN_TEXT || "100";
+const CONFIG_BACKEND_URL =
+  process.env.CONFIG_BACKEND_URL || "https://api.rudderlabs.com";
+
+const functionListCache = new NodeCache();
+const FUNC_LIST_KEY = "fn-list";
+
+const delayInMs = async (ms = 2000) =>
+  new Promise(resolve => setTimeout(resolve, ms));
+
+const callWithRetry = async (fn, count = 0, ...args) => {
+  try {
+    return await fn(...args);
+  } catch (err) {
+    if (count > 2) {
+      throw err;
+    }
+    await delayInMs();
+    return callWithRetry(fn, count + 1);
+  }
+};
+
+const isFunctionDeployed = functionName => {
+  const funcList = functionListCache.get(FUNC_LIST_KEY) || [];
+  return funcList.includes(functionName);
+};
+
+const setFunctionInCache = functionName => {
+  const funcList = functionListCache.get(FUNC_LIST_KEY) || [];
+  funcList.push(functionName);
+  functionListCache.set(FUNC_LIST_KEY, funcList);
+};
+
+const deployFaasFunction = async (functionName, code, versionId, testMode) => {
+  try {
+    logger.debug("[Faas] Deploying a faas function");
+    let envProcess = "python index.py";
+
+    if (!testMode) {
+      envProcess = `${envProcess} --vid ${versionId} --config-backend-url ${CONFIG_BACKEND_URL}`;
+    } else {
+      envProcess = `${envProcess} --code "${code}"`;
+    }
+    // TODO: investigate and add more required labels and annotations
+    const payload = {
+      service: functionName,
+      name: functionName,
+      image: FAAS_BASE_IMG,
+      envProcess,
+      labels: {
+        faas_function: functionName,
+        "com.openfaas.scale.max": FAAS_MAX_PODS_IN_TEXT,
+        "parent-component": "openfaas"
+      },
+      annotations: {
+        "prometheus.io.scrape": "true"
+      }
+    };
+
+    await deployFunction(payload);
+    logger.debug("[Faas] Deployed a faas function");
+  } catch (error) {
+    logger.error(
+      `[Faas] Error while deploying ${functionName}: ${error.message}`
+    );
+    // To handle concurrent create requests,
+    // throw retry error if already exists so that request can be retried
+    if (error.message.includes("already exists")) {
```

Comment:
Is there a code we can rely on in this case, instead of an error message? As error messages tend to change.

Reply:
for given image: `ghcr.io/openfaas/faas-netes:0.15.4` this error message would not change

Reply:
if there is a change in error message, integration tests would fail